Readme.rst
TQueues
=======

TQueues (reThinkdb Queues) is a simple python3.5+
library for queueing jobs and processing them in workers using rethinkdb.


Features
--------

- Accepts *coroutines* as jobs
- Distributed
- Handles everything in a *rethinkdb database object* that can be reused
  at the worker
- Exposes changes via websockets
- Uses rethinkdb streaming "changes" method, wich is pretty efficient
- Uses asyncio, implements an async context manager for jobs and
  async iterator for the workers

::

                            +--> Worker
                            |
    Rethindb --> Dispatcher ---> Worker
                            |
                            +--> Worker




Usage
=====

TQueues provides two entry points, tqueues_dispatcher and tqueues_worker.

TQueues dispatcher must be accesible from all workers, and you need to start
one worker per parallel task you need. They may be distributed (as long as
they can reach the dispatcher)

TQueues worker
--------------

::

    Usage:
    	tqueues_worker -h | --help
    	tqueues_worker --version
    	tqueues_worker --endpoint_url <endpoint_url>
    	tqueues_worker --queue <queue>

    Options:
    	-h --help                        Show this screen
    	-v --version                     Show version
    	--endpoint_url <ENDPOINT_URL>    TQueues dispatcher endpoint
    	--queue        <QUEUE>           Endpoint queue to listen on

    Examples:
    	tqueues_worker --endpoint_url http://127.0.0.1:800/ --queue testqueue


TQueues job dispatcher
----------------------

::

    Usage:
        tqueues_dispatcher --db "db" --host "127.0.0.1" --port 28015
        tqueues_dispatcher --db "db" --host "127.0.0.1"
        tqueues_dispatcher --db "db" --port 28015
        tqueues_dispatcher --db "db" --user 'user'
        tqueues_dispatcher --db "db" --password 'password'
        tqueues_dispatcher -h | --help
        tqueues_dispatcher --version

    Options:
        --host "127.0.0.1"                          Rethinkdb host
        --db "db"                                   Rethinkdb databaes
        --port 28015                                Rethinkdb port
        --user 'user'                               Rethinkdb user
        --password 'password'                       Rethinkdb password
        --allowed_domains 'foo.com,bar.com'         Allowed domains
        --loglevel (DEBUG|INFO)                     Loglevel
        -h   --help                                 Show this screen
        --version                                   Show version

    Examples:
        tqueues_dispatcher --host localhost --db foo --port 28015 --user foo --password bar --loglevel INFO --allowed_domains 'foo.com,bar.com'


ChangeLog
CHANGES
=======

0.1.7
-----

* fixed readthedocs
* Travis fix

0.1.6
-----

* Improved docopt usage

0.1.5
-----

* Doc improvements (closes: #1)
* Added basic rethinkdb connection capabilities on the workers
* typo
* Added endpoints
* Reorganized code and fixed #6
* Made a better sample function
* Removed filter, it wouldn't want to work
* Removed the need to send a startup message via websocket, yet still seems to fail to monitor for changes
* Fixed delete and updated. Added websocket handler
* Fixed lastests commits
* Added cors, due to that, moved config to a conffile
* ups
* Trying environment and readthedocs

0.1.4
-----

* Docs
* initial documentation

0.1.3
-----

* Updates
* Revert pbr
* Trying to build only 3.5
* That was from other project =P
* Empty dispatcher tests TODO
* Travis ci
* Added development requirements
* Added documentation
* When a queue does not exist and we launch multiple workers against it we could find a race condition on the table creation. As it does not really affect anything at all, we just ignore the exception
* Worker now handles table creation
* I like it better this way, job.work uses job's context manager, so deletion of the task is handled inside job itself
* Handle job deletion in the tests
* Tests for Worker class
* Added basic test for Worker()
* Added a basic test function to be used in tests. Not ideal, but fast
* It seems to be a bit clearer this way. This was implemented as an async context manager wich was an async iterator... but that, with a job queue, is not as clear as it could be
* Better syntax
* Added delete method
* We cant get an eventloop inside an eventloop
* Sleep two seconds each time the server gives a non-200 response
* Stop deleting stuff from the database, make it stateful. Also, rewrote workers so that dispatchers are themselves, wich looks cool

0.1.2
-----

* Now we can execute coroutines

0.1.0
-----

* version bump
* emptyreadme
* Readme rename
* Readme rename -a
* cfg
* Queue creation is now up the client, also implemented changes' include_initial=True
* GPL
* Rename
* Added empty readme
* gitignore
* Deleted .eggs
* Initial commit


test/unit/conftest.py
import pytest
import asyncio
from aiohttp import web


@pytest.yield_fixture
def create_server(event_loop):
    """ Returns a server"""
    app = handler = srv = None
    loop = event_loop

    @asyncio.coroutine
    def create(*, debug=False, ssl_ctx=None, proto='http'):
        nonlocal app, handler, srv
        app = web.Application(loop=loop)
        port = 8000
        handler = app.make_handler(debug=debug, keep_alive_on=False)
        srv = yield from loop.create_server(handler, '127.0.0.1', port,
                                            ssl=ssl_ctx)
        if ssl_ctx:
            proto += 's'
        url = "{}://127.0.0.1:{}".format(proto, port)
        return app, url

    yield create

    @asyncio.coroutine
    def finish():
        yield from handler.finish_connections()
        yield from app.finish()
        srv.close()
        yield from srv.wait_closed()

    loop.run_until_complete(finish())


test/unit/test_tqueues_basic.py
import pytest
from aiohttp import web


class TestWorker:
    @pytest.mark.asyncio
    @pytest.mark.parametrize('id', list(range(10)))
    async def test_single_task(self, create_server, id):
        from tqueues import Worker

        self.deleted = False

        def get_response(id):
            """ Get response json """
            return {
                'queue': 'test',
                'args': [id],
                'kwargs': {'param': id},
                'method': 'tqueues.test',
                'id': id
            }

        response = get_response(id)

        async def handler(request):
            """
                Return always test task
            """
            return web.json_response(response)

        async def handler_del(request):
            """
                DELETED stuff
            """
            self.deleted = True
            return web.Response(body=b'OK')

        app, url = await create_server()
        app.router.add_route('GET', '/', handler)
        app.router.add_route('DELETE', '/', handler_del)
        worker = Worker(url, "test")

        async for job in worker:
            result = await job.work()
            job = result.pop('tq_parent_job')
            assert result == tuple([tuple(response['args']),
                                    response['kwargs']])
            assert self.deleted
            break

    @pytest.mark.asyncio
    @pytest.mark.parametrize('id', list(range(10)))
    async def test_multi_task(self, create_server, id):
        from tqueues import Worker

        self.deleted = False

        def get_response():
            """ Get response json """
            for id in range(10):
                yield {
                    'queue': 'test',
                    'args': [id],
                    'kwargs': {'param': id},
                    'method': 'tqueues.test',
                    'id': id
                }

        response = get_response()

        async def handler(request):
            """
                Return always test task
            """
            self.current_response = next(response)
            return web.json_response(self.current_response)

        async def handler_del(request):
            """
                DELETED stuff
            """
            self.deleted = True
            return web.Response(body=b'OK')

        app, url = await create_server()
        app.router.add_route('GET', '/', handler)
        app.router.add_route('DELETE', '/', handler_del)
        worker = Worker(url, "test")

        enum = 0
        async for job in worker:
            enum += 1
            res = await job.work()
            assert res == tuple([tuple(self.current_response['args']),
                                 self.current_response['kwargs']])
            assert self.deleted
            if enum == 9:
                break

    @pytest.mark.asyncio
    @pytest.mark.parametrize('id', list(range(10)))
    async def test_queue_creation(self, create_server, id):
        from tqueues import Worker

        self.deleted = False
        self.created = False

        def get_response():
            """ Get response json """
            for id in range(10):
                yield {
                    'queue': id,
                    'args': [id],
                    'kwargs': {'param': id},
                    'method': 'tqueues.test',
                    'id': id
                }

        response = get_response()

        async def handler(request):
            """
                Return always test task
            """
            raise web.HTTPNotImplemented()

        async def handler_put(request):
            """ PUT stuff """
            self.created = True
            return web.Response(body=b'OK')

        self.current_response = next(response)

        app, url = await create_server()
        app.router.add_route('GET', '/', handler)
        app.router.add_route('PUT', '/', handler_put)
        worker = Worker(url, self.current_response)

        enum = 0
        async for job in worker:
            enum += 1
            job
            assert self.created
            if enum == 9:
                break


test/unit/test_dispatcher_basic.py
def test_create_task():
    pass


def test_delete_task():
    pass


def test_consume_task():
    pass


tqueues/worker.py
"""

TQueues worker

Usage:
    tqueues_worker -h | --help
    tqueues_worker --version
    tqueues_worker --endpoint_url <endpoint_url>
    tqueues_worker --queue <queue>

Options:
    --endpoint_url <ENDPOINT_URL>    TQueues dispatcher endpoint
    --host         <host>            Rethinkdb host
    --db           <db>              Rethinkdb databaes
    --port         <port>            Rethinkdb port
    --user         <user>            Rethinkdb user
    --password     <password>        Rethinkdb password
    --queue        <QUEUE>           Endpoint queue to listen on
    -h --help                        Show this screen
    -v --version                     Show version

Examples:
    tqueues_worker --endpoint_url http://127.0.0.1:800/ --queue testqueue

"""

from docopt import docopt
import rethinkdb as r
import aiohttp
import asyncio
import inspect
import importlib


def client():
    """
        Client
    """
    opts = docopt(__doc__, version="0.0.1")
    endpoint_url = False

    with suppress(KeyError):
        endpoint_url = opts.pop('endpoint_url')
    queue = opts.pop('queue')
    rethinkdb = opts

    loop = asyncio.get_event_loop()
    runner = Worker(endpoint_url, rethinkdb, queue).run_forever()
    loop.run_until_complete(runner)
    return loop.close()


class Job:
    """
        Context manager that handles a job on the dispatcher.

        This executes the given method (see `Job.method`) adding
        itself to it via "tq_parent_job" keyword argument.
        That means your worker methods need to accept a tq_parent_job.

        That makes the job able to update information in the rethinkdb
        using the update method of tq_parent_job.

    """
    def __init__(self, endpoint_url, rethinkdb, data):
        self.endpoint_url = endpoint_url
        self.rethinkdb = rethinkdb
        self.data = data

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, value, tback):
        with aiohttp.ClientSession() as session:
            queue = self.data['queue']
            await session.delete(self.endpoint_url, params={
                'id': self.data["id"], 'queue': queue})

    async def update(self, data):
        """
            Update remote object in the database providing `data`
            If rethinkdb connection parameters are specified, they're
            used instead of the patch in the endpoint.

            .. warning:: If you specify different rethinkdb parameters
                         here than the used on the endpoint, it'll lead to
                         unexpected behavior
        """
        if not self.rethinkdb:
            with aiohttp.ClientSession() as session:
                queue = self.data['queue']
                return await session.patch(self.endpoint_url, params={
                    "queue": queue, 'id': self.data['id']}, data=data)
        else:
            opts = self.rethinkdb
            conn = await r.connect(**opts)
            queue = r.db(self.rethinkdb['db']).table(self.queue)
            return await queue.get(self.request.GET['id']).update(data).run(
                conn)

    @property
    def method(self):
        """
            Finds and imports the method specified in an importable format
            (foo.bar:baz would use baz from foo.bar)
        """
        module, method = self.data['method'].rsplit(':', 1)
        method = getattr(importlib.import_module(module), method)
        return method

    async def work(self):
        """
            Run the job

            .. warning:: Job MUST accept a named parameter ``job``
            containing this object.
        """
        async with self:
            try:
                args, kwargs = self.data['args'], self.data['kwargs']
                kwargs.update({'tq_parent_job': self})
                if inspect.iscoroutinefunction(self.method):
                    return await self.method(*args, **kwargs)
                return self.method(*args, **kwargs)
            except:
                return False


class Worker:
    """
        Implements a simple worker over the http api
    """

    def __init__(self, endpoint_url, rethinkdb, queue):
        self.endpoint_url = endpoint_url
        self.rethinkdb = rethinkdb
        self._id = False
        self.queue = queue

    async def __aiter__(self):
        return self

    async def __anext__(self):
        try:
            with aiohttp.ClientSession() as session:
                args = (self.endpoint_url,)
                kwargs = {"params": {'queue': self.queue}}
                async with session.get(*args, **kwargs) as resp:
                    if resp.status == 501:
                        async with session.put(*args, **kwargs) as resp:
                            assert resp.status == 200
                    assert resp.status == 200
                    data = await resp.json()
                    return Job(self.endpoint_url, self.rethinkdb, data)
        except AssertionError:
            await asyncio.sleep(2)

    async def run_forever(self):
        """
            We iterate over the job dispatcher object
            assigned to us.
        """
        async for job in self:
            if job:
                await job.work()


async def test(*args, **kwargs):
    """ Test function for worker"""
    import datetime
    curr_data = kwargs['tq_parent_job'].data.copy()
    curr_data.update({'date_start':
                      datetime.datetime.today().strftime('%D %H:%M')})
    await kwargs['tq_parent_job'].update(curr_data)
    await asyncio.sleep(5)
    curr_data.update({'foo': 'bar'})
    await kwargs['tq_parent_job'].update(curr_data)
    return args, kwargs


tqueues/dispatcher.py
#!/usr/bin/env python3.5
"""
Tqueues job dispatcher.

Usage:
    tqueues_dispatcher --db "db" --queue "queue" --host "127.0.0.1"
    tqueues_dispatcher --db "db" --queue "queue" --port 28015
    tqueues_dispatcher --db "db" --queue "queue" --user 'user'
    tqueues_dispatcher --db "db" --queue "queue" --password 'password'
    tqueues_dispatcher --db "db" --queue "queue" --loglevel INFO
    tqueues_dispatcher --db "db" --queue "queue"
    tqueues_dispatcher -h | --help
    tqueues_dispatcher --version

Options:
    --host         <host>                       Rethinkdb host
    --db           <db>                         Rethinkdb databaes
    --port         <port>                       Rethinkdb port
    --user         <user>                       Rethinkdb user
    --password     <password>                   Rethinkdb password
    --allowed_domains 'foo.com,bar.com'         Allowed domains
    --loglevel (DEBUG|INFO)                     Loglevel
    -h   --help                                 Show this screen
    --version                                   Show version

Examples:
    tqueues_dispatcher --host localhost --db foo --port 28015 --user foo \
--password bar --loglevel INFO --allowed_domains 'foo.com,bar.com'

"""

from docopt import docopt
import logging
import json
from contextlib import suppress
import aiohttp_cors
from aiohttp import web
import rethinkdb as r


TASK_PENDING = 'pending'
TASK_FINISHED = 'finished'
TASK_STARTED = 'started'


class Dispatcher(web.View):
    """
        Exposes an API for the workers that run as simple consumers
        When a worker finishes its task, consumes the next one.

        Accepts ``rethinkdb.connect`` args.
        For a more secure environment, you should ONLY HAVE ONE DISPATCHER
        for database in rethinkdb.
    """

    async def get(self):
        """
        .. http:get:: /?queue={string:queue}

            Gets a task from the dispatcher.

            If we run into a race condition, it'll return a
            404 for the client to retry

        **Example request**:

        .. sourcecode:: http

            GET /?queue=foo
            Host: example.com
            Accept: application/json, text/javascript

        **Example response**:

        .. sourcecode:: http

            HTTP/1.1 200 OK
            Vary: Accept
            Content-Type: text/javascript
            {
              'queue': 'foo', method': 'method.to.execute',
              'args': (*args), 'kwargs': {**kwargs}
            }

        .. note::

            method.to.execute must be a method importable in the workers' side

        :query queue: queue (table) where to listen in the database
        :statuscode 200: no error
        :statuscode 404: Race condition happened, task is no longer present
        """
        opts = self.request.app['rethinkdb']
        conn = await r.connect(**opts)
        queue = r.db(opts['db']).table(self.request.GET['queue'])
        filter_ = queue.filter({'status': TASK_PENDING})

        cursor = await filter_.changes(include_initial=True).run(conn)
        await cursor.fetch_next()
        result = await cursor.next()

        try:
            new_id = result['new_val']['id']
            dres = await queue.get(new_id).run(conn)
            assert dres['status'] != TASK_FINISHED
            assert dres['status'] != TASK_STARTED
            dres = await queue.get(new_id).update(
                {'status': TASK_STARTED}).run(conn)
            assert dres['replaced'] != 0
        except AssertionError:
            logging.debug('Task is being consumed already')
            raise web.HTTPNotFound()

        return web.Response(text=json.dumps(result['new_val']))

    async def post(self):
        """
            Creates a new task, just dumps whatever
            we have to the database after a brief format check

              {
                'queue': 'foo', method': 'method.to.execute',
                'args': (args), 'kwargs': {kwargs}
              }

        .. http:post:: /

            Gets a task from the dispatcher.

            If the table does not exist, it returns a 501 for
            the client to handle it

        **Example request**:

        .. sourcecode:: http

            POST /
            Host: example.com
            Accept: application/json, text/javascript

            {
              'queue': 'foo', method': 'method.to.execute',
              'args': (*args), 'kwargs': {**kwargs}
            }


        **Example response**:

        .. sourcecode:: http

            HTTP/1.1 200 OK
            Vary: Accept
            Content-Type: text/javascript

            ok

        .. note::

            method.to.execute must be a method importable in the workers' side

        :<json string queue: Queue (table) to add this task to
        :<json array args: List of positional arguments to pass to method
        :<json array kwargs: List of keyword arguments to pass to method
        :<json string method: Method to import and execute

        :statuscode 200: No error
        :statuscode 501: Table does not exist
        :statuscode 400: Not all params have been specified
        :statuscode 404: No more tasks in the queue, retry later

        """
        opts = self.request.app['rethinkdb']
        conn = await r.connect(**opts)
        await self.request.post()

        mandatory = ['queue', 'method']
        if not all([a in self.request.POST for a in mandatory]):
            raise web.HTTPBadRequest()

        queue = r.db(opts['db']).table(self.request.POST['queue'])
        data = dict(self.request.POST)
        if 'args' not in data:
            data['args'] = []
        if 'kwargs' not in data:
            data['kwargs'] = {}
        data.update({'status': TASK_PENDING})

        try:
            id_ = await queue.insert(data).run(conn)
        except r.errors.ReqlOpFailedError as err:
            if 'does not exist.' in err.message:
                raise web.HTTPNotImplemented()
            raise web.HTTPNotFound('No more tasks')
        data.update({'id': id_['generated_keys'][0]})
        return web.json_response(data)

    async def put(self):
        """
        .. http:put:: /?queue={string:queue}

            Creates a queue if it does not exist.

        **Example request**:

        .. sourcecode:: http

            GET /?queue=foo
            Host: example.com
            Accept: application/json, text/javascript

        **Example response**:

        .. sourcecode:: http

            HTTP/1.1 200 OK
            Vary: Accept
            Content-Type: text/javascript

            ok

        :query queue: queue (table) to create
        :statuscode 200: This method always should return 200

        """
        opts = self.request.app['rethinkdb']
        conn = await r.connect(**opts)
        qname = self.request.GET['queue']
        with suppress(r.errors.ReqlOpFailedError):
            r.db(opts['db']).table_create(qname).run(conn)

        return web.Response(body=b'ok')

    async def delete(self):
        """
        .. http:delete:: /?queue={string:queue}&id={string:id}

        Marks a task as completed

        **Example request**:

        .. sourcecode:: http

            DELETE /?id=foo&queue=bar
            Host: example.com
            Accept: application/json, text/javascript

        **Example response**:

        .. sourcecode:: http

            HTTP/1.1 200 OK
            Vary: Accept
            Content-Type: text/javascript

            ok

        :query id: id to mark as completed
        :query queue: queue (table) to work on
        :statuscode 200: This method always should return 200


        """
        opts = self.request.app['rethinkdb']
        conn = await r.connect(**opts)
        queue = r.db(opts['db']).table(self.request.GET['queue'])
        res = await queue.get(self.request.GET['id']).update(
            {'status': TASK_FINISHED}).run(conn)
        assert res['replaced'] == 1
        return web.json_response(res)

    async def patch(self):
        """

            In case the job is unable to access directly rethinkdb, we
            can use this endpoint to update the data.
            See `Job` documentation on `Job.update`.
        """
        opts = self.request.app['rethinkdb']
        conn = await r.connect(**opts)
        queue = r.db(opts['db']).table(self.request.GET['queue'])
        data = await self.request.post()
        res = await queue.get(self.request.GET['id']).update(data).run(conn)
        assert res['replaced'] == 1
        return web.json_response(res)


async def wshandle(request):
    """
        Websocket handler
        Right now this only waits on given queue and dumps
        'new_val' on rethinkdb changes.

        .. TODO:: #5
    """
    ws_ = web.WebSocketResponse()
    await ws_.prepare(request)

    opts = request.app['rethinkdb']
    conn = await r.connect(**opts)
    queue = r.db(opts['db']).table(request.GET['queue'])
    cursor = await queue.changes(include_initial=True).run(conn)

    while cursor.fetch_next():
        ws_.send_str(json.dumps(dict(await cursor.next())['new_val']))

    return ws_


def server():
    """
        Starts main dispatch server
    """
    allowed_domains = []

    opts = docopt(__doc__, version="0.0.1")

    logging.basicConfig(level=getattr(logging, opts.pop('loglevel')))
    r.set_loop_type("asyncio")

    with suppress(KeyError):
        allowed_domains = opts.pop('allowed_domains').split(',')
        allowed_domains = [a.strip() for a in allowed_domains]

    app = web.Application()
    app['rethinkdb'] = opts

    default_opts = aiohttp_cors.ResourceOptions(
        allow_credentials=True, expose_headers="*", allow_headers="*")
    cors = aiohttp_cors.setup(app, defaults={
        dom: default_opts for dom in allowed_domains})

    cors.add(cors.add(app.router.add_resource('/')).add_route('*', Dispatcher))
    cors.add(cors.add(app.router.add_resource('/ws')).add_route('*', wshandle))
    web.run_app(app)
